lib/Core/Executor.cpp

```diff
@@ -2548,6 +2548,159 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
     // instructions.
     terminateStateOnExecError(state, "Unexpected ShuffleVector instruction");
     break;
+
+  case Instruction::AtomicRMW: {
+    // An atomic instruction gets a pointer and a value. It reads the value at the pointer, perfoms its operation, stores the result and returns the value that was originally at the pointer.
+    const AtomicRMWInst* ai = cast<AtomicRMWInst>(i);
+    switch (ai->getOperation()) {
+    case AtomicRMWInst::Xchg: {
+      ref<Expr> pointer = eval(ki, 0, state).value;
```

Comment:
* This is a really minor benefit, that most IDEs do for you anyway. A simple `sed` script can do the same job too.

* I would consider this a feature. If you only have to deal with is `Expr` and don't care what actual type it is.  What happens if suddenly `.value` only starts returning `AddExpr`, someone might think that you can write logic here where `.value` is an AddExpr, which is obiosuly not the case here. With explicit `ref<Expr>` you communicate the intention that the lowest denominator for all `Expr`s here is `Expr`.

* Yes, that's good? Because it reminds you this location exists?

You are also completely ignoring the value of someone unfamiliar with this code looking at this declaration and having no clue what type pointer is with looking inside eval. And having to rely on IDEs is not an excuse, there are plenty of cases where you don't use an IDE for example Github.

Reply:
* nope, `sed` will replace all of the `ref<Expr>` on a file base not a more specific granularity; I'm not sure about IDEs either that change the variable type based on the return type of a function.
* nope, if you want `ref<Expr>` you should communicating it by explicitly casting it or explicitly using that type, currently there is no need for this restriction at this type.
* nope, if you really have to change it, the compiler will tell you. otherwise you force to many things.

Unfamiliar problem? Type helps you least.

Reply:
Type helps me quite a lot. Also using `auto` and than casting is a bad solution in my opinion, becaue it just clutters your screen with `auto`, which is basically saying I refuse to give any information.

I don't see any point in arguing further. Someone has to make a call ...

Reply:
It's the other way around, you cast and put the solution in an auto variable,
like `auto foo = cast<Expr>(bar)`.

Reply:
which has `auto` and `cast` and `<>` cluttering your screen, how is it better than
`Expr foo = bar`